timelock-corruption-detection/src/main/java/com/palantir/timelock/history/PaxosLogHistoryProgressTracker.java

```diff
@@ -0,0 +1,112 @@
+ * limitations under the License.
+ */
+
+package com.palantir.timelock.history;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.palantir.paxos.Client;
+import com.palantir.paxos.NamespaceAndUseCase;
+import com.palantir.timelock.history.models.LearnerUseCase;
+import com.palantir.timelock.history.models.ProgressComponents;
+import com.palantir.timelock.history.models.SequenceBounds;
+import com.palantir.timelock.history.sqlite.LogVerificationProgressState;
+import com.palantir.timelock.history.sqlite.SqlitePaxosStateLogHistory;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import javax.sql.DataSource;
+
+public class PaxosLogHistoryProgressTracker {
+    public static final int MAX_ROWS_ALLOWED = 500;
+
+    private final LogVerificationProgressState logVerificationProgressState;
+    private final SqlitePaxosStateLogHistory sqlitePaxosStateLogHistory;
+
+    private Map<NamespaceAndUseCase, ProgressComponents> verificationProgressStateCache = new ConcurrentHashMap<>();
+
+    public PaxosLogHistoryProgressTracker(
+            DataSource dataSource, SqlitePaxosStateLogHistory sqlitePaxosStateLogHistory) {
+        this.logVerificationProgressState = LogVerificationProgressState.create(dataSource);
+        this.sqlitePaxosStateLogHistory = sqlitePaxosStateLogHistory;
+    }
+
+    public SequenceBounds getPaxosLogSequenceBounds(NamespaceAndUseCase namespaceAndUseCase) {
+        ProgressComponents progress = getOrPopulateProgressComponents(namespaceAndUseCase);
+        return SequenceBounds.builder()
+                .lower(progress.seq())
+                .upper(progress.seq() + MAX_ROWS_ALLOWED)
+                .build();
+    }
+
+    public void updateProgressState(Map<NamespaceAndUseCase, SequenceBounds> namespaceAndUseCaseSequenceBoundsMap) {
+        namespaceAndUseCaseSequenceBoundsMap.forEach((namespaceAndUseCase, bounds) ->
+                updateProgressStateForNamespaceAndUseCase(namespaceAndUseCase, bounds));
+    }
+
+    private ProgressComponents getOrPopulateProgressComponents(NamespaceAndUseCase namespaceAndUseCase) {
+        return verificationProgressStateCache.computeIfAbsent(namespaceAndUseCase, this::getLastVerifiedSeqFromLogs);
+    }
+
+    private ProgressComponents getLastVerifiedSeqFromLogs(NamespaceAndUseCase namespaceAndUseCase) {
+        Client client = namespaceAndUseCase.namespace();
+        String useCase = namespaceAndUseCase.useCase();
+
+        return logVerificationProgressState
+                .getProgressComponents(client, useCase)
+                .orElseGet(() -> logVerificationProgressState.resetProgressState(
+                        client, useCase, getLatestSequenceForNamespaceAndUseCase(namespaceAndUseCase)));
+    }
+
+    @VisibleForTesting
+    void updateProgressStateForNamespaceAndUseCase(NamespaceAndUseCase key, SequenceBounds value) {
+        long lastVerifiedSequence = value.upper();
```

Comment:
This is an exclusive bound, so it is _not_ the last verified sequence (see comment on my thoughts on SequenceBound).